libparsec/crates/platform_mountpoint/src/windows/mod.rs

```diff
@@ -0,0 +1,481 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+mod error;
+
+use anyhow::anyhow;
+use once_cell::sync::{Lazy, OnceCell};
+use std::{
+    ops::Deref,
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+};
+
+use winfsp_wrs::{
+    filetime_from_utc, filetime_now, u16cstr, CleanupFlags, CreateFileInfo, CreateOptions,
+    FileAccessRights, FileAttributes, FileInfo, FileSystem, FileSystemContext, PSecurityDescriptor,
+    Params, SecurityDescriptor, U16CStr, U16CString, VolumeInfo, VolumeParams,
+    WriteMode as WinWriteMode, NTSTATUS, STATUS_DIRECTORY_NOT_EMPTY, STATUS_NOT_IMPLEMENTED,
+    STATUS_RESOURCEMANAGER_READ_ONLY,
+};
+
+use libparsec_types::FileDescriptor;
+
+use crate::{EntryInfo, EntryInfoType, MountpointInterface, MountpointManager, WriteMode};
+
+static SD: Lazy<SecurityDescriptor> = Lazy::new(|| {
+    SecurityDescriptor::from_wstr(u16cstr!("O:BAG:BAD:P(A;;FA;;;SY)(A;;FA;;;BA)(A;;FA;;;WD)"))
+        .expect("unreachable, valid SecurityDescriptor")
+});
+
+static VOLUME: OnceCell<Arc<Mutex<VolumeInfo>>> = OnceCell::new();
+
+const ENTRY_INFO_EXTENSION: &str = "__parsec_entry_info__";
+const MAX_FILE_NODES: u64 = 1024;
+const MAX_FILE_SIZE: u64 = 16 * 1024 * 1024;
+const ALLOCATION_UNIT: u64 = 4096;
+const FILE_NODES: u64 = 1;
+const VOLUME_LABEL: &U16CStr = u16cstr!("parsec");
+
+fn is_entry_info_path(path: &Path) -> bool {
+    path.extension()
+        .map(|ext| ext == ENTRY_INFO_EXTENSION)
+        .unwrap_or_default()
+}
+
+impl From<&EntryInfo> for FileInfo {
+    fn from(value: &EntryInfo) -> Self {
+        let created = filetime_from_utc(value.created);
+        let updated = filetime_from_utc(value.updated);
+
+        match value.ty {
+            EntryInfoType::File => *FileInfo::default()
+                .set_file_attributes(value.ty.into())
+                .set_creation_time(created)
+                .set_last_access_time(updated)
+                .set_last_write_time(updated)
+                .set_change_time(updated)
+                .set_index_number(value.id.as_u128() as u64)
+                .set_file_size(value.size)
+                .set_allocation_size(
+                    ((value.size + ALLOCATION_UNIT - 1) / ALLOCATION_UNIT) * ALLOCATION_UNIT,
```

Comment:
From [winfsp tutorial](https://winfsp.dev/doc/WinFsp-Tutorial/):

> Files in a Windows file system can have two sizes: an "EndOfFile" size or FileSize and an AllocationSize. The FileSize is the number of bytes contained in a file. The AllocationSize is a concept that **many file systems can safely ignore** (or not expose to the kernel): it is the actual number of bytes that a file occupies on its storage medium.
> 
> [...]
> Regardless of whether a file system exposes AllocationSize it must obey the following rule: **it must always be that FileSize <= AllocationSize**.

In Parsec we never allocate more space for the file data, so this allocation size is meaningless.
I think we should just do `set_allocation_size(value.size)` and provide a comment stating 
```rust
// AllocationSize is the size actually occupied on the storage medium, this is
// meaningless for Parsec (however WinFSP requires FileSize <= AllocationSize)
```

Reply:
There is one test which will fail then in `winfsp_tests`

Reply:
nvm